// jsonc-eslint-parser/lib/parser/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseForESLint = void 0;
const espree_1 = require("./espree");
const errors_1 = require("./errors");
const visitor_keys_1 = require("./visitor-keys");
const convert_1 = require("./convert");
const token_store_1 = require("./token-store");
function parseForESLint(code, options) {
    const parserOptions = Object.assign({ filePath: "<input>", ecmaVersion: 2019 }, options || {}, {
        loc: true,
        range: true,
        raw: true,
        tokens: true,
        comment: true,
        eslintVisitorKeys: true,
        eslintScopeManager: true,
    });
    try {
        const ast = parseJS(`0(${code}\n)`, parserOptions);
        const tokens = ast.tokens || [];
        const tokenStore = new token_store_1.TokenStore(tokens);
        const statement = ast.body[0];
        const callExpression = statement.expression;
        const expression = callExpression.arguments[0];
        if (!expression) {
            return errors_1.throwEmptyError("an expression");
        }
        if (expression && expression.type === "SpreadElement") {
            return errors_1.throwUnexpectedTokenError("...", expression);
        }
        if (callExpression.arguments[1]) {
            const node = callExpression.arguments[1];
            return errors_1.throwUnexpectedTokenError(",", tokenStore.getTokenBefore(node, token_store_1.isComma));
        }
        tokens.shift();
        tokens.shift();
        tokens.pop();
        const last = tokens[tokens.length - 1];
        if (last && token_store_1.isComma(last)) {
            return errors_1.throwUnexpectedTokenError(",", last);
        }
        ast.range[1] = statement.range[1] = last.range[1];
        ast.loc.end.line = statement.loc.end.line = last.loc.end.line;
        ast.loc.end.column = statement.loc.end.column = last.loc.end.column;
        ast.body = [statement];
        statement.expression = expression;
        return {
            ast: postprocess(ast, tokenStore, parserOptions),
            visitorKeys: visitor_keys_1.KEYS,
            services: {
                isJSON: true,
            },
        };
    }
    catch (err) {
        return errors_1.throwErrorAsAdjustingOutsideOfCode(err, code);
    }
}
exports.parseForESLint = parseForESLint;
function parseJS(code, options) {
    const espree = espree_1.getEspree();
    try {
        return espree.parse(code, options);
    }
    catch (err) {
        const perr = errors_1.ParseError.normalize(err);
        if (perr) {
            convert_1.fixErrorLocation(perr);
            throw perr;
        }
        throw err;
    }
}
function postprocess(ast, tokenStore, options) {
    const ctx = getJSONSyntaxContext(options === null || options === void 0 ? void 0 : options.jsonSyntax);
    const jsonAst = convert_1.convertNode(ast, tokenStore, ctx);
    const tokens = [];
    for (const token of ast.tokens || []) {
        tokens.push(convert_1.convertToken(token));
    }
    const comments = ast.comments || [];
    if (!ctx.comments && comments.length > 0) {
        return errors_1.throwUnexpectedCommentError(comments[0]);
    }
    for (const comment of comments) {
        convert_1.fixLocation(comment);
    }
    jsonAst.tokens = tokens;
    jsonAst.comments = comments;
    return jsonAst;
}
function getJSONSyntaxContext(str) {
    const upperCase = str === null || str === void 0 ? void 0 : str.toUpperCase();
    if (upperCase === "JSON") {
        return {
            trailingCommas: false,
            comments: false,
            plusSigns: false,
            spacedSigns: false,
            leadingOrTrailingDecimalPoints: false,
            infinities: false,
            nans: false,
            numericSeparators: false,
            invalidJsonNumbers: false,
            multilineStrings: false,
            unquoteProperties: false,
            singleQuotes: false,
            numberProperties: false,
            undefinedKeywords: false,
            sparseArrays: false,
            regExpLiterals: false,
            templateLiterals: false,
            bigintLiterals: false,
        };
    }
    if (upperCase === "JSONC") {
        return {
            trailingCommas: true,
            comments: true,
            plusSigns: false,
            spacedSigns: false,
            leadingOrTrailingDecimalPoints: false,
            infinities: false,
            nans: false,
            numericSeparators: false,
            invalidJsonNumbers: false,
            multilineStrings: false,
            unquoteProperties: false,
            singleQuotes: false,
            numberProperties: false,
            undefinedKeywords: false,
            sparseArrays: false,
            regExpLiterals: false,
            templateLiterals: false,
            bigintLiterals: false,
        };
    }
    if (upperCase === "JSON5") {
        return {
            trailingCommas: true,
            comments: true,
            plusSigns: true,
            spacedSigns: true,
            leadingOrTrailingDecimalPoints: true,
            infinities: true,
            nans: true,
            numericSeparators: false,
            invalidJsonNumbers: true,
            multilineStrings: true,
            unquoteProperties: true,
            singleQuotes: true,
            numberProperties: false,
            undefinedKeywords: false,
            sparseArrays: false,
            regExpLiterals: false,
            templateLiterals: false,
            bigintLiterals: false,
        };
    }
    return {
        trailingCommas: true,
        comments: true,
        plusSigns: true,
        spacedSigns: true,
        leadingOrTrailingDecimalPoints: true,
        infinities: true,
        nans: true,
        numericSeparators: true,
        invalidJsonNumbers: true,
        multilineStrings: true,
        unquoteProperties: true,
        singleQuotes: true,
        numberProperties: true,
        undefinedKeywords: true,
        sparseArrays: true,
        regExpLiterals: true,
        templateLiterals: true,
        bigintLiterals: true,
    };
}
